src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.hpp

```diff
@@ -0,0 +1,488 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <boost/functional/hash.hpp>  // IWYU pragma: keep
+#include <cstddef>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tags.hpp"       // IWYU pragma: keep
+#include "DataStructures/Variables.hpp"  // IWYU pragma: keep
+#include "Domain/Direction.hpp"
+#include "Domain/DirectionMap.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoGridHelpers.hpp"
+#include "NumericalAlgorithms/LinearOperators/ApplyMatrices.hpp"
+#include "NumericalAlgorithms/LinearOperators/MeanValue.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/EqualWithinRoundoff.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+template <size_t VolumeDim>
+class Element;
+template <size_t VolumeDim>
+class ElementId;
+template <size_t>
+class Mesh;
+/// \endcond
+
+namespace SlopeLimiters {
+namespace Hweno_detail {
+
+// Caching class that holds various precomputed terms used in the constrained-
+// fit algebra on each element.
+//
+// In principle, the algebra depends in detail on the number, size, and mesh of
+// the neighbors in each directions. However, by restricting to grids with no
+// h/p-refinement, we get crucial simplifications:
+// 1. The element has at most one neighbor per dimension.
+// 2. The mesh on every neighbor is the same as on the element.
+//
+// With these restrictions, the number of independent configurations to be
+// cached is greatly reduced. In particular, any elements that share the same
+// configuration of internal/external boundaries vs. direction can share the
+// same cached data. Because there are 2*VolumeDim boundaries, there are
+// 2^(2*VolumeDim) possible configurations of internal/external boundaries, and
+// therefore 2^(2*VolumeDim) configurations to be cached.
```

Comment:
Just so I understand, does this counting include the combinatorics associated with excluded/primary directions, or are those choices extra factors on this storage cost?

Reply:
These are extra factors. I tried to clarify the wording.